partiql-parser/src/scanner.rs

```diff
@@ -0,0 +1,270 @@
+struct Remainder<'val> {
+    /// The remainder of text to scan.
+    input: &'val str,
+    /// The offset in the input translate locations from
+    offset: LineAndColumn,
+}
+
+impl<'val> Remainder<'val> {
+    /// Produces a new [`Remainder`] by slicing the input by the given amount and providing
+    /// a new line/column baseline.
+    ///
+    /// The offset given is the logical one as if the input was the start of the sequence.
+    /// This method will calculate the new [`Remainder`] offset based on that.
+    fn consume(&self, amount: usize, offset: LineAndColumn) -> Self {
+        Self {
+            input: &self.input[amount..],
+            offset: self.offset.move_relative(offset),
+        }
+    }
+}
+
+/// A lexeme of PartiQL derived from a slice of input text.
+///
+/// A token can be thought of as a sort of continuation, it knows where it is in the input list
+/// and allows scanning to resume from it.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Token<'val> {
+    /// The underlying value of the token.
+    content: Content<'val>,
+    /// Start location for this token.
+    start_location: LineAndColumn,
+    /// End location for this token.
+    end_location: LineAndColumn,
+    /// Slice from the input containing the whole token.
+    text: &'val str,
+    /// The remaining unscanned value after this token.
+    remainder: Remainder<'val>,
+}
+
+impl<'val> Token<'val> {
+    /// Returns the parsed content of this token.
+    pub fn content(&self) -> Content<'val> {
+        self.content
+    }
+
+    /// Returns the location of where this token starts in the input.
+    pub fn start_loc(&self) -> LineAndColumn {
+        self.start_location
+    }
+
+    /// Returns the location of where this token ends in the input.
+    pub fn end_loc(&self) -> LineAndColumn {
+        self.end_location
+    }
+
+    /// Returns the slice of the input encompassing the entire token.
+    pub fn text(&self) -> &str {
+        self.text
+    }
+
+    /// Returns the remainder of the input after this token.
+    pub fn rem(&self) -> &str {
+        self.remainder.input
+    }
+}
+
+/// Returns tokens from a given slice of input text.
+pub trait Scanner<'val> {
+    /// Returns the next token from the input.
+    fn next_token(&mut self) -> ParserResult<Token<'val>>;
+}
+
+/// Root scanner, leverages the underlying PEG to provide the scanner.
+pub struct PartiQLScanner<'val> {
+    /// The remaining input to scan.
+    remainder: Remainder<'val>,
+}
+
+impl<'val> PartiQLScanner<'val> {
+    fn do_next_token(&mut self) -> ParserResult<Token<'val>> {
+        // the scanner rule is expected to return a single node
+        let pair: Pair<'val, Rule> =
+            PartiQLParser::parse(Rule::Scanner, self.remainder.input)?.exactly_one()?;
+        let start_location = self.remainder.offset.move_relative(pair.start_loc());
+        let end_location = self.remainder.offset.move_relative(pair.end_loc());
```

Comment:
Notice here we translate the parser which has a logical line/column from its start, to where we left off.